File: src/main/java/com/airsaid/localization/translate/impl/google/GoogleToken.java
/*
 * Copyright 2021 Airsaid. https://github.com/airsaid
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.airsaid.localization.translate.impl.google;

import com.airsaid.localization.translate.util.AgentUtil;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.util.Pair;
import com.intellij.util.io.HttpRequests;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * @author airsaid
 */
public class GoogleToken {

  private static final Logger LOG = Logger.getInstance(GoogleToken.class);

  private static final int MIM = 60 * 60 * 1000;
  private static final Random GENERATOR = new Random();
  private static final Pattern TKK_PATTERN = Pattern.compile("tkk='(\\d+).(-?\\d+)'");
  private static final String ELEMENT_URL = "%s/translate_a/element.js";

  private static Pair<Long, Long> sInnerValue = Pair.create(0L, 0L);
  private static boolean sNeedUpdate = true;

  public static String getToken(String text) {
    return getToken(text, getDefaultTKK());
  }

  public static String getToken(String text, Pair<Long, Long> tkk) {
    int length = text.length();
    List<Long> a = new ArrayList<>();
    int b = 0;
    char[] ch = text.toCharArray();
    while (b < length) {
      int c = ch[b];
      if (128 > c) {
        a.add((long) c);
      } else {
        if (2048 > c) {
          a.add((long) (c >> 6 | 192));
        } else {
          if (55296 == (c & 64512) && b + 1 < length && 56320 == (ch[b + 1] & 64512)) {
            c = 65536 + ((c & 1023) << 10) + (ch[++b] & 1023);
            a.add((long) (c >> 18 | 240));
            a.add((long) (c >> 12 & 63 | 128));
          } else {
            a.add((long) (c >> 12 | 224));
          }
          a.add((long) (c >> 6 & 63 | 128));
        }
        a.add((long) (c & 63 | 128));
      }
      b++;
    }

    long d = tkk.first;
    long e = tkk.second;
    long f = d;
    for (Long h : a) {
      f += h;
      f = fun(f, "+-a^+6");
    }

    f = fun(f, "+-3^+b+-f");
    f = f ^ e;
    if (0 > f) {
      f = (f & Integer.MAX_VALUE) + Integer.MAX_VALUE + 1;
    }
    f = (long) (f % 1E6);

    return f + "." + (f ^ d);
  }

  private static Long fun(Long a, String b) {
    long g = a;
    char[] ch = b.toCharArray();
    for (int c = 0; c < ch.length - 1; c += 3) {
      char d = ch[c + 2];
      int e = 'a' <= d ? (d - 87) : d - '0';
      long f = '+' == ch[c + 1] ? g >>> e : g << e;
      g = '+' == ch[c] ? g + f & ((long) Integer.MAX_VALUE * 2 + 1) : g ^ f;
    }
    return g;
  }

  private static Pair<Long, Long> getDefaultTKK() {
    long now = System.currentTimeMillis() / MIM;
    long curVal = sInnerValue.first;
    if (!sNeedUpdate && now == curVal) {
      return sInnerValue;
    }

    Pair<Long, Long> newTKK = getTKKFromGoogle();
    sNeedUpdate = newTKK == null;
    sInnerValue = newTKK != null ? newTKK : Pair.create(now, Math.abs((long) GENERATOR.nextInt()) + (long) GENERATOR.nextInt());

    return sInnerValue;
  }

  private static Pair<Long, Long> getTKKFromGoogle() {
    try {
      String url = String.format(ELEMENT_URL, GoogleTranslator.HOST_URL);
      LOG.info("getTKKFromGoogle url: " + url);
      String elementJs = HttpRequests.request(url)
          .userAgent(AgentUtil.getUserAgent())
          .tuner(connection -> connection.setRequestProperty("Referer", GoogleTranslator.HOST_URL))
          .readString();
      Matcher matcher = TKK_PATTERN.matcher(elementJs);
      if (matcher.find()) {
        long value1 = Long.parseLong(matcher.group(1));
        long value2 = Long.parseLong(matcher.group(2));
        LOG.info(String.format("TKK: %d.%d", value1, value2));
        return Pair.create(value1, value1);
      }
    } catch (Exception e) {
      e.printStackTrace();
      LOG.warn("TKK get failed.", e);
    }
    return null;
  }

}

File: src/main/kotlin/com/airsaid/localization/translate/impl/google/GoogleToken.kt
package com.airsaid.localization.translate.impl.google

import com.airsaid.localization.translate.util.HttpRequestFactory
import com.intellij.openapi.diagnostic.Logger
import com.intellij.util.concurrency.annotations.RequiresBackgroundThread
import java.lang.StrictMath.abs
import java.util.concurrent.ThreadLocalRandom
import java.util.regex.Pattern

private const val TOKEN_TTL_MILLIS = 60 * 60 * 1000L
private const val ELEMENT_JS_PATH = "/translate_a/element.js"

private val TKK_PATTERN: Pattern = Pattern.compile("tkk='(\\d+).(-?\\d+)'")

private data class Token(val value1: Long, val value2: Long, val hour: Long)

private val LOG: Logger = Logger.getInstance("GoogleToken")

private var cachedToken: Token? = null
private val tokenLock = Any()

/**
 * Clears any cached TKK tokens so subsequent calls fetch fresh credentials.
 *
 * @author airsaid
 */
internal fun resetGoogleTokenCache() {
  synchronized(tokenLock) {
    cachedToken = null
  }
}

/**
 * Returns the current TKK token pair, fetching or regenerating when required.
 *
 * @author airsaid
 */
@RequiresBackgroundThread
internal fun currentTkk(): Pair<Long, Long> {
  synchronized(tokenLock) {
    val nowHour = System.currentTimeMillis() / TOKEN_TTL_MILLIS
    cachedToken?.takeIf { it.hour == nowHour }?.let { return it.value1 to it.value2 }

    val updated = fetchFromGoogle()?.takeIf { it.hour == nowHour } ?: generateLocal(nowHour)
    cachedToken = updated
    return updated.value1 to updated.value2
  }
}

private fun fetchFromGoogle(): Token? {
  return try {
    val url = googleApiUrl(ELEMENT_JS_PATH)
    val response = HttpRequestFactory.get(url, 10_000)
      .withGoogleHeaders()
      .connect { it.readString(null) }
    val matcher = TKK_PATTERN.matcher(response)
    if (!matcher.find()) {
      LOG.warn("TKK not found in element.js response")
      null
    } else {
      val nowHour = System.currentTimeMillis() / TOKEN_TTL_MILLIS
      val value1 = matcher.group(1).toLong()
      val value2 = matcher.group(2).toLong()
      Token(value1, value2, nowHour)
    }
  } catch (error: Throwable) {
    LOG.warn("Failed to refresh Google TKK", error)
    null
  }
}

private fun generateLocal(hour: Long): Token {
  val random = ThreadLocalRandom.current()
  val value = abs(random.nextInt().toLong()) + random.nextInt().toLong()
  return Token(hour, value, hour)
}

/**
 * Computes the Google translate token for the given payload.
 *
 * @author airsaid
 */
internal fun String.tk(): String {
  val (d, e) = currentTkk()
  val bytes = mutableListOf<Long>()
  var index = 0
  while (index < length) {
    var charCode = this[index].code
    when {
      charCode < 0x80 -> bytes += charCode.toLong()
      charCode < 0x800 -> {
        bytes += (charCode shr 6 or 0xC0).toLong()
        bytes += (charCode and 0x3F or 0x80).toLong()
      }

      charCode in 0xD800..0xDBFF && index + 1 < length -> {
        val next = this[index + 1].code
        if (next and 0xFC00 == 0xDC00) {
          charCode = 0x10000 + ((charCode and 0x3FF) shl 10) + (next and 0x3FF)
          bytes += (charCode shr 18 or 0xF0).toLong()
          bytes += (charCode shr 12 and 0x3F or 0x80).toLong()
          bytes += (charCode shr 6 and 0x3F or 0x80).toLong()
          bytes += (charCode and 0x3F or 0x80).toLong()
          index++
        } else {
          bytes += (charCode shr 12 or 0xE0).toLong()
          bytes += (charCode shr 6 and 0x3F or 0x80).toLong()
          bytes += (charCode and 0x3F or 0x80).toLong()
        }
      }

      else -> {
        bytes += (charCode shr 12 or 0xE0).toLong()
        bytes += (charCode shr 6 and 0x3F or 0x80).toLong()
        bytes += (charCode and 0x3F or 0x80).toLong()
      }
    }
    index++
  }

  var result = d
  for (byte in bytes) {
    result += byte
    result = applyTransformation(result, "+-a^+6")
  }
  result = applyTransformation(result, "+-3^+b+-f")
  result = result xor e
  if (result < 0) {
    result = (result and 0x7FFFFFFF) + 0x80000000 + 1
  }
  result %= 1_000_000

  return "$result.${result xor d}"
}

private fun applyTransformation(value: Long, op: String): Long {
  var acc = value
  var i = 0
  while (i < op.length - 2) {
    val shiftChar = op[i + 2]
    val shift = if (shiftChar >= 'a') shiftChar.code - 87 else shiftChar.digitToInt()
    val operand = if (op[i + 1] == '+') acc ushr shift else acc shl shift
    acc = if (op[i] == '+') (acc + operand) and 0xFFFFFFFFL else acc xor operand
    i += 3
  }
  return acc
}
